File: src/org/lineageos/xiaomi_bluetooth/mma/MMAResponse.java
package org.lineageos.xiaomi_bluetooth.mma;

import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.util.Arrays;


public class MMAResponse {

    public static String TAG = MMAResponse.class.getName();
    public static boolean DEBUG = true;

    public final byte opCode;
    public final byte opCodeSN;
    public final byte[] data;

    protected MMAResponse(byte opCode, byte opCodeSN, byte[] data) {
        this.opCode = opCode;
        this.opCodeSN = opCodeSN;
        this.data = data;
    }

    @NonNull
    @Override
    public String toString() {
        return "MMAResponse{" +
                "opCode=" + opCode +
                ", opCodeSN=" + opCodeSN +
                ", data=" + Arrays.toString(data) +
                '}';
    }

    @Nullable
    public static MMAResponse fromPacket(byte[] packet) {
        if (packet == null || packet.length < 6) {
            if (DEBUG) Log.d(TAG, "fromPacket: packet length < 6");
            return null;
        }

        int type = (packet[0] & 0x80) == 0 ? 0 : 1;
        int responseFlag = (packet[0] & 0x40) == 0 ? 0 : 1;
        byte opCode = packet[1];
        int parameterLength = ((packet[2] & 0xFF) << 8) | (packet[3] & 0xFF);
        byte status = packet[4];
        byte opCodeSN = packet[5];

        if (type != 0 || responseFlag != 0) {
            if (DEBUG) Log.d(TAG, "fromPacket: type or responseFlag not 0");
//            return null;
        }
        if (parameterLength < 2 || packet.length - 4 != parameterLength) {
            if (DEBUG) Log.d(TAG, "fromPacket: parameterLength not equal");
            return null;
        }
        if (status != 0) {
            if (DEBUG) Log.d(TAG, "fromPacket: status not success");
        }

        byte[] data = new byte[parameterLength - 2];
        System.arraycopy(packet, 6, data, 0, parameterLength - 2);

        if (DEBUG) Log.d(TAG, "fromPacket: opCode " + opCode + " opCodeSN " + opCodeSN);
        return new MMAResponse(opCode, opCodeSN, data);
    }

}

File: src/org/lineageos/xiaomi_tws/mma/MMAResponse.kt
package org.lineageos.xiaomi_tws.mma

import android.bluetooth.BluetoothDevice
import android.util.Log
import org.lineageos.xiaomi_tws.EarbudsConstants.XIAOMI_MMA_RESPONSE_STATUS_OK
import org.lineageos.xiaomi_tws.utils.ByteUtils.bytesToInt

open class MMAResponse protected constructor(
    val device: BluetoothDevice,
    val type: Int,
    val flag: Int,
    val opCode: Byte,
    val opCodeSN: Byte,
    val status: Byte,
    val data: ByteArray
) {

    val ok = status == XIAOMI_MMA_RESPONSE_STATUS_OK

    override fun toString(): String {
        return "MMAResponse{device=$device, opCode=$opCode, opCodeSN=$opCodeSN, status=$status, data=${data.contentToString()}}"
    }

    companion object {
        private val TAG = MMAResponse::class.java.simpleName
        private const val DEBUG = true

        fun fromPacket(device: BluetoothDevice, packet: ByteArray): MMAResponse {
            check(packet.size >= 6) { "packet length < 6" }

            val type = if ((packet[0].toInt() and 0x80) == 0) 0 else 1
            val flag = if ((packet[0].toInt() and 0x40) == 0) 0 else 1
            val opCode = packet[1]
            val parameterLength = bytesToInt(packet[2], packet[3])
            val status = packet[4]
            val opCodeSN = packet[5]

            if (type != 0 || flag != 0) {
                if (DEBUG) Log.d(
                    TAG,
                    "Response type or flag not 0: type $type, responseFlag: $flag"
                )
            }
            check(parameterLength >= 2 && packet.size - 4 == parameterLength) {
                "Parameter length not equal, expect $parameterLength, bug ${packet.size - 4}"
            }
            if (status.toInt() != 0) {
                Log.w(TAG, "fromPacket: status not success")
            }

            val data = ByteArray(parameterLength - 2).apply {
                System.arraycopy(packet, 6, this, 0, parameterLength - 2)
            }

            return MMAResponse(device, type, flag, opCode, opCodeSN, status, data).also {
                if (DEBUG) Log.d(TAG, "fromPacket: $it")
            }
        }
    }
}
